src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java

```diff
@@ -377,120 +376,157 @@ private String help(MetricRule rule, String unit, String statistic) {
+
+      for (List<Dimension> dimensions : getDimensions(rule, config.client)) {
+        queries.addAll(queriesForRuleDimensions(rule, dimensions));
+      }
+
+      // GetMetricData will not accept more than 100 queries per request
+      int numQueriesPerRequest = 100;
+      List<List<MetricDataQuery>> queryChunks = new ArrayList<>();
+      for (int i=0; i < queries.size(); i += numQueriesPerRequest) {
+          queryChunks.add(queries.subList(i, Math.min(i + numQueriesPerRequest, queries.size())));
+      }
+
+      for (List<MetricDataQuery> chunk : queryChunks) {
+        GetMetricDataRequest request = new GetMetricDataRequest()
+                .withStartTime(startDate)
+                .withEndTime(endDate)
+                .withScanBy(ScanBy.TimestampDescending);
+
+        request.setMetricDataQueries(chunk);
+        requests.add(request);
+      }
+
+      return requests;
+    }
+
+  private Map<String, MetricDataResult> fetchResults(ActiveConfig config, long start) {
+    Map<String, MetricDataResult> idsToResults = new HashMap<>();
+    for (MetricRule rule : config.rules) {
+      final List<GetMetricDataRequest> getMetricDataRequests = requestForRule(rule, config, start);
+      for (GetMetricDataRequest request : getMetricDataRequests) {
+        if (request == null) {
+        } else {
+          String nextToken = null;
+          do {
+            request.setNextToken(nextToken);
+            cloudwatchRequests.labels("getMetricData", rule.awsNamespace).inc();
+            GetMetricDataResult getResult = config.client.getMetricData(request);
+            for (MetricDataResult result : getResult.getMetricDataResults()) {
+              idsToResults.put(result.getId(), result);
+            }
+            nextToken = getResult.getNextToken();
+          } while (nextToken != null);
         }
+      }
+    }
+    return idsToResults;
+  }
+
+  private void scrape(List<MetricFamilySamples> mfs) throws CloneNotSupportedException {
+    ActiveConfig config = (ActiveConfig) activeConfig.clone();
+
+    long start = System.currentTimeMillis();
+
+    Map<String, MetricDataResult> idsToResults = fetchResults(config, start);
+
+    for (MetricRule rule : config.rules) {
+      Map<String, List<MetricFamilySamples.Sample>> statsTypeToSamples =
+              new HashMap<String, List<MetricFamilySamples.Sample>>();
 
-        for (List<Dimension> dimensions: getDimensions(rule, config.client)) {
-          request.setDimensions(dimensions);
+      String jobName = safeName(rule.awsNamespace.toLowerCase());
+      String baseName = safeName(rule.awsNamespace.toLowerCase() + "_" + toSnakeCase(rule.awsMetricName));
+
+      if (rule.awsNamespace.equals("AWS/DynamoDB")
+              && rule.awsDimensions.contains("GlobalSecondaryIndexName")
+              && brokenDynamoMetrics.contains(rule.awsMetricName)) {
+        baseName += "_index";
+      }
+
+      for (List<Dimension> dimensions : getDimensions(rule, config.client)) {
```

Comment:
getDimensions makes API calls, which were already made above. We shouldn't be calling it twice.